### PR TITLE
[spinel] save MAC frame counter to a local variable

### PR DESCRIPTION
**Pull request by zhanglongxia**
This commit saves the MAC counter to a local variable to avoid calling the core function `Get<Settings>().Read()` in `radio_spinel_impl.hpp`. This is useful when spinel module is used as a lib in other projects.

### CODE REVIEW

File: src/lib/spinel/radio_spinel_impl.hpp [hunk 1]
```diff
@@ -2145,13 +2154,9 @@ template <typename InterfaceType> void RadioSpinel<InterfaceType>::RestoreProper
...
 
-    if (mInstance != nullptr)
+    if (mMacFrameCounterSet)
     {
-        if (static_cast<Instance *>(mInstance)->template Get<Settings>().Read(networkInfo) == OT_ERROR_NONE)
-        {
-            SuccessOrDie(
-                Set(SPINEL_PROP_RCP_MAC_FRAME_COUNTER, SPINEL_DATATYPE_UINT32_S, networkInfo.GetMacFrameCounter()));
-        }
+        SuccessOrDie(Set(SPINEL_PROP_RCP_MAC_FRAME_COUNTER, SPINEL_DATATYPE_UINT32_S, mMacFrameCounter));
```
Comment (abtink) [inline src/lib/spinel/radio_spinel_impl.hpp:2159]: There is still some small chance that radio/RCP has used some counter after `mMacFrameCounter` (for enh ack) and they are in queue to be sent to host (not yet processed by host `RadioSpinel`).

So how about having `mMacFrameCounter` track the last used value we see from RCP? And here we add some guard jump when we set it, maybe setting it to `mMacFrameCounter + 10` (jump over 10?).

File: src/lib/spinel/radio_spinel_impl.hpp [hunk 2]
```diff
@@ -1743,6 +1746,14 @@ void RadioSpinel<InterfaceType>::HandleTransmitDone(uint32_t          aCommand,
         VerifyOrExit(unpacked > 0, error = OT_ERROR_PARSE);
         static_cast<Mac::TxFrame *>(mTransmitFrame)->SetKeyId(keyId);
         static_cast<Mac::TxFrame *>(mTransmitFrame)->SetFrameCounter(frameCounter);
+
+#if OPENTHREAD_SPINEL_CONFIG_RCP_RESTORATION_MAX_COUNT > 0
+        mMacFrameCounterSet = true;
+        // If we directly save the `frameCounter`, the RCP will use the `frameCounter` to send the next frame after the
+        // RCP is restored. The peer will consider a duplicate frame received and then drop the frame. Here saves the
+        // `frameCounter + 1` to avoid this case from happening.
+        mMacFrameCounter = frameCounter + 1;
```
Comment (abtink) [inline src/lib/spinel/radio_spinel_impl.hpp:1755]: Should we also track/update the counter used for enhanced ack? I think we see them in `ParseRadioFrame()` as `aFrame.mInfo.mRxInfo.mAckFrameCounter`?

**Review by abtink — COMMENTED** [review]
Thanks @zhanglongxia .

Two suggestions below:

File: src/lib/spinel/radio_spinel_impl.hpp (hunk 1 above)
Comment (zhanglongxia) [reply src/lib/spinel/radio_spinel_impl.hpp]: I have updated the code to add the guard.

File: src/lib/spinel/radio_spinel_impl.hpp (hunk 2 above)
Comment (zhanglongxia) [reply src/lib/spinel/radio_spinel_impl.hpp]: I have updated the code to track the counter for enhanced ACK.

File: src/lib/spinel/radio_spinel_impl.hpp [hunk 3]
```diff
@@ -2145,13 +2156,15 @@ template <typename InterfaceType> void RadioSpinel<InterfaceType>::RestoreProper
...
 
-    if (mInstance != nullptr)
+    if (mMacFrameCounterSet)
     {
-        if (static_cast<Instance *>(mInstance)->template Get<Settings>().Read(networkInfo) == OT_ERROR_NONE)
-        {
-            SuccessOrDie(
-                Set(SPINEL_PROP_RCP_MAC_FRAME_COUNTER, SPINEL_DATATYPE_UINT32_S, networkInfo.GetMacFrameCounter()));
-        }
+        const uint8_t kFrameCounterGuard = 10;
```
Comment (abtink) [inline src/lib/spinel/radio_spinel_impl.hpp:2161]:
```suggestion
       static constexpr uint8_t kFrameCounterGuard = 10;
```

**Review by abtink — APPROVED** [review]
Looks great. Thanks. 👍

File: src/lib/spinel/radio_spinel_impl.hpp (hunk 3 above)
Comment (zhanglongxia) [reply src/lib/spinel/radio_spinel_impl.hpp]: Updated

File: src/lib/spinel/radio_spinel_impl.hpp [hunk 4]
```diff
@@ -2145,13 +2159,15 @@ template <typename InterfaceType> void RadioSpinel<InterfaceType>::RestoreProper
...
 
-    if (mInstance != nullptr)
+    if (mMacFrameCounterSet)
     {
-        if (static_cast<Instance *>(mInstance)->template Get<Settings>().Read(networkInfo) == OT_ERROR_NONE)
-        {
-            SuccessOrDie(
-                Set(SPINEL_PROP_RCP_MAC_FRAME_COUNTER, SPINEL_DATATYPE_UINT32_S, networkInfo.GetMacFrameCounter()));
-        }
+        static constexpr uint8_t kFrameCounterGuard = 10;
```
Comment (jwhui) [inline src/lib/spinel/radio_spinel_impl.hpp:2164]: Why 10 vs. some other value? Can we provide some reasoning as to how this value was picked?

Comment (abtink) [reply src/lib/spinel/radio_spinel_impl.hpp:2164]: I suggested this in https://github.com/openthread/openthread/pull/9407#discussion_r1318693325.

For any use of frame counter by RCP (such as for enhanced acks that may not yet be reported up to the host and are in the queue). I suggested 10 as an example. We can consider using a larger value to be safe.

We can try to determine a practical worst-case (upper estimate) delay from an action on RCP (e.g., enhanced ack send to a received frame) till its related Spinel frame is received by the host and is processed, taking into account upper estimates for interface, queuing, and processing delays. We can then consider how many such events may occur within this window, i.e., how many frames RCP can practically receive and send enhanced acks for (which would consume a frame counter) within this upper-estimate window. 

Thoughts?

Comment (jwhui) [reply src/lib/spinel/radio_spinel_impl.hpp:2164]: Yes, I think we should at least document in a comment the rationale for the chosen value, including any data / assumptions that were used.

Comment (zhanglongxia) [reply src/lib/spinel/radio_spinel_impl.hpp]: I have updated the frame counter guard value and added some comments to explain why we chose this value.

**Review by abtink — APPROVED** [review]
LGTM. Thanks. 👍

**Review by jwhui — APPROVED** [review]

**Pull request merged**